### ISSUE TITLE
[Bug] feline causes the neovim splash screen to disappear

### ISSUE THREAD

**Issue by ibhagwan** [famiu/feline.nvim#23]
**Neovim version**
Both 0.6 nightly and 0.5 release:
```
NVIM v0.6.0-dev+63-ga8c3d50fa
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
```
```
NVIM v0.5.0
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
```

**Feline version**
`4c04c967691d028b4056f726735db329039cdff2`

This was happening since the first version of `feline` I used, this isn't a new issue.

**Describe the bug**
Feline causes the neovim spash screen to disappear, for clarity I'm referring to the opening screen with the neovim version, etc:
```
~                     NVIM v0.5.0
~   
~      Nvim is open source and freely distributable
~                 https://neovim.io/#chat
~   
~     type  :help nvim<Enter>       if you are new!
~     type  :checkhealth<Enter>     to optimize Nvim
~     type  :q<Enter>               to exit
~     type  :help<Enter>            for help
~   
~              Help poor children in Uganda!
~     type  :help iccf<Enter>       for information
```

**To Reproduce**

Run feline as the only plugin and it should reproduce, it even happens when lazy loading to with packer on `VimEnter`:
```lua
use { 'famiu/feline.nvim',
  requires = { 'kyazdani42/nvim-web-devicons' },
  config = "require'plugin.feline'",
  event = 'VimEnter'
}
```

**Expected behavior**
Keep the splash screen.

**Does this error occur in the minimal init file?**

Yes, the issue is reproducible with `minimal_init.lua`.

**Additional context**

Most fancy status line plugins I tried have the same issue, for exmaple [galaxyline issue #107](https://github.com/glepnir/galaxyline.nvim/issues/107), `gitsigns` used to have a similar problem, perhaps [gitsigns issue #238](https://github.com/lewis6991/gitsigns.nvim/issues/238) will provide a clue how to fix it?

**Comment by famiu** [comment]
As of now I have no idea what could be causing that to happen. I might look into it further when I'm not busy but in  the meantime, if anyone knows what could be causing it, it'd be great if they could inform me of it

**Comment by ibhagwan** [comment]
> As of now I have no idea what could be causing that to happen. I might look into it further when I'm not busy but in the meantime, if anyone knows what could be causing it, it'd be great if they could inform me of it

Ok, I've done some troubleshooting, the problematic code is in [generator.lua#L76](https://github.com/famiu/feline.nvim/blob/master/lua/feline/generator.lua#L76) which is the `cmd(string.format('highlight %s gui=%s guifg=%s guibg=%s', hlname, style, fg, bg))` command.

In an attempt to solve it I tried scheduling the highlight for later, I may have done it incorrectly, I'm not sure if  running the `async:send` right after schedule would achieve the goal here...  here's the piece of code I changed, any ideas?

```lua
        local async
        async = vim.loop.new_async(vim.schedule_wrap(function()
          cmd(string.format('highlight %s gui=%s guifg=%s guibg=%s', hlname, style, fg, bg))
          async:close()
        end))
        async:send()
        highlights[hlname] = true
        return hlname
```

**Comment by ibhagwan** [comment]
[More info on the same issue from lualine](https://github.com/hoob3rt/lualine.nvim/issues/259)

**Comment by famiu** [comment]
> [More info on the same issue from lualine](https://github.com/hoob3rt/lualine.nvim/issues/259)

It seems it's caused by asynchronous loading of the statusline in the case of Lualine, but Feline doesn't load anything asynchronously, so that shouldn't be what's causing it

**Comment by ibhagwan** [comment]
> > [More info on the same issue from lualine](https://github.com/hoob3rt/lualine.nvim/issues/259)
> 
> It seems it's caused by asynchronous loading of the statusline in the case of Lualine, but Feline doesn't load anything asynchronously, so that shouldn't be what's causing it

It might not be exactly the same issue but it’s def coming from the `highlight` command

**Comment by ibhagwan** [comment]
Closing this issue, [details in my PR with the fix](https://github.com/famiu/feline.nvim/pull/24).

**Comment by ram02z** [comment]
This should be reopened, the bug still exists. The issue doesn't exist in shadmansaleh's lualine.

**Comment by famiu** [comment]
> This should be reopened, the bug still exists. The issue doesn't exist in shadmansaleh's lualine.

I'm not exactly sure what's causing the issue honestly. And the "fix" provided in the linked PR seemed to break the statusline for others. I wonder what Lualine is doing differently.

**Comment by ram02z** [comment]
> 
> I'm not exactly sure what's causing the issue honestly. And the "fix" provided in the linked PR seemed to break the statusline for others. I wonder what Lualine is doing differently.

The fix in the lualine fork was to remove the async code on startup, but feline doesn't use any async code on startup and `vim.schedule` doesn't seem to be causing it either.

**Comment by famiu** [comment]
> > I'm not exactly sure what's causing the issue honestly. And the "fix" provided in the linked PR seemed to break the statusline for others. I wonder what Lualine is doing differently.
> 
> The fix in the lualine fork was to remove the async code on startup, but feline doesn't use any async code on startup and `vim.schedule` doesn't seem to be causing it either.

Maybe it's coming from the `update_inactive_windows` function? Just a hunch.

**Comment by ram02z** [comment]
> 
> Maybe it's coming from the `update_inactive_windows` function? Just a hunch.

It's not, I disabled the autocommands since they both use `update_inactive_windows` and I can still reproduce.

**Comment by famiu** [comment]
Then it's likely to be caused by defining highlights after startup, in which case it'd be a bug in Neovim itself.

**Comment by ibhagwan** [comment]
FYI, I’m still running [my own fork of feline](https://github.com/ibhagwan/feline.nvim) due to this fix, the latest changes to the generator made it a tiny bit different but the essence still remains, [here is the commit with the fix](https://github.com/ibhagwan/feline.nvim/commit/51c7420b068014e6a89e993cd736f5ec048e5c94).

>Then it's likely to be caused by defining highlights after startup, in which case it'd be a bug in Neovim itself.

I’m not sure if it’s a bug or an obscure behavior but the chances of it being prioritized within neovim are not high, I already did the work and told you exactly the reason why this happens, if my “fix” doesn’t work for others you can still fix it yourself by creating the highlight groups before the status line generator is called so that the highlights aren’t created as part of the status line generation.

@ram02z, can you try [my fork](https://github.com/ibhagwan/feline.nvim) and let @famiu know if this solves it for you?

**Comment by famiu** [comment]
> FYI, I’m still running [my own fork of feline](https://github.com/ibhagwan/feline.nvim) due to this fix, the latest changes to the generator made it a tiny bit different but the essence still remains, [here is the commit with the fix](https://github.com/ibhagwan/feline.nvim/commit/51c7420b068014e6a89e993cd736f5ec048e5c94).
> 
> > Then it's likely to be caused by defining highlights after startup, in which case it'd be a bug in Neovim itself.
> 
> I’m not sure if it’s a bug or an obscure behavior but the chances of it being prioritized within neovim are not high

Still worth reporting it nonetheless. If you never report it it'll never be fixed.

> I already did the work and told you exactly the reason why this happens, if my “fix” doesn’t work for others you can still fix it yourself by creating the highlight groups before the status line generator is called so that the highlights aren’t created as part of the status line generation.

That's impossible since `hl` can be a function which can give different outputs.

> @ram02z, can you try [my fork](https://github.com/ibhagwan/feline.nvim) and let @famiu know if this solves it for you?

Even if it does work, it still seems to break highlights for others. And honestly, if I have to choose, I'd prefer startup screen disappearing due to a Neovim bug over all highlights disappearing due to a Feline bug. Though if possible I'd opt for neither.

**Comment by ibhagwan** [comment]
> That's impossible since hl can be a function which can give different outputs.

You can do that once just for the startup and then go back to dynamic generation, is it a big hassle just for the splash screen? Yes :)

> Even if it does work, it still seems to break highlights for others. And honestly, if I have to choose, I'd prefer startup screen disappearing due to a Neovim bug over all statusline highlights disappearing due to a Feline bug. Though if possible I'd opt for neither.

I understand, that’s why I’m running my own fork of the fix and continue to rebase against upstream.

Also, is it possible since your generator changes it doesn’t break the code for others anymore?

**Comment by famiu** [comment]
> Also, is it possible since your generator changes it doesn’t break the code for others anymore?

That very well could be the case. But I'm not sure how to test that anymore without taking a risk and breaking it for others again, which I'd prefer avoiding

**Comment by ram02z** [comment]
I haven't looked into how the generator works in regards to highlights, but I cannot seem to get a minimal reproduction to create an issue upstream. The following code doesn't remove the splash screen.
```lua
local function add_hl(name, fg, bg, style)
    vim.api.nvim_command(string.format(
        'highlight %s gui=%s guifg=%s guibg=%s',
        name,
        style,
        fg,
        bg
    ))
end

function M.statusline(winid)
    add_hl('StatusComponentDefault', '#FFFFFF', '#000000', 'NONE')
    return '%#StatusComponentDefault#%f'
end
```

**Comment by famiu** [comment]
Did #74 fix your issue?

**Comment by famiu** [comment]
closing the issue since it was fixed

**Comment by ram02z** [comment]
39aafe1aab75e2bacd7471f1a95c5b30870d2562 brings this issue back and I am going crazy trying to debug it. Using `colors.bg` works but when using `colors.fg`. it doesn't work.

**Comment by famiu** [comment]
> [9685971](https://github.com/famiu/feline.nvim/commit/9685971de476a4468a01d00f3d6eee527bad61d5) brings this issue back and I am going crazy trying to debug it. Using `colors.bg` works but when using `colors.fg`. it doesn't work.

the splash screen works fine for me though

**Comment by ram02z** [comment]
> 
> the splash screen works fine for me though. maybe it has something to do with your config?

hmm, I am using the default config to test.

**Comment by famiu** [comment]
> > the splash screen works fine for me though. maybe it has something to do with your config?
> 
> hmm, I am using the default config to test.

Oh.. I'm using a different config. After trying the default config again I can now reproduce the issue

**Comment by ram02z** [comment]
> 
> Oh.. I'm using a different config. After trying the default config again I can now reproduce the issue. So that confirms that it's happening due to the config itself, specifically due to how the config handles the fg and bg values

Do you you have any seperators with highlights in your config? When I was testing, that was causing the splash screen to not show.

**Comment by famiu** [comment]
> > Oh.. I'm using a different config. After trying the default config again I can now reproduce the issue. So that confirms that it's happening due to the config itself, specifically due to how the config handles the fg and bg values
> 
> Do you have any seperators with highlights in your config? When I was testing, that was causing the splash screen to not show.

Yes, multiple of them

**Comment by ram02z** [comment]
Alright, I probably didn't test properly. Can you share your config?

**Comment by famiu** [comment]
Here: https://gitlab.com/famiu/dot-nvim/-/blob/master/lua/config/ui/feline.lua

**Comment by ram02z** [comment]
@famiu Honestly, not sure where to start with debugging this, but still 9685971 is incorrect. `parse_sep` has the comment `By default, foreground color of separator is background color of parent`, yet you changed it to inherit colors.fg instead of colors.bg. For some reason, using `colors.fg` causes this issue on the default config, but `colors.bg` doesn't.

**Comment by famiu** [comment]
> @famiu Honestly, not sure where to start with debugging this, but still [9685971](https://github.com/famiu/feline.nvim/commit/9685971de476a4468a01d00f3d6eee527bad61d5) is incorrect. `parse_sep` has the comment `By default, foreground color of separator is background color of parent`, yet you changed it to inherit colors.fg instead of colors.bg. For some reason, using `colors.fg` causes this issue on the default config, but `colors.bg` doesn't.

The issue with inheriting `colors.bg` is that `colors.bg` is also used for the background of the separator by default, so then the background and foreground will become the same

**Comment by famiu** [comment]
For me, the issue still occurs under the latest `develop`, but only very specific conditions, which are:

1. The provider is `git_branch`
2. It has a `right_sep` component value as in the `default` preset
3. The directory that Neovim is started from is a git repo

All 3 of these conditions need to be satisfied to reproduce the bug, if any 1 of these is different, the bug doesn't occur, which is very intriguing. @ibhagwan and @ram02z, thoughts?

**Comment by ram02z** [comment]
@famiu I am assuming you can reproduce the issue when using the default config. I have tried to recreate the scenario with the conditions you listed and unforunately, I still cannot reproduce.

**Comment by famiu** [comment]
> @famiu I am assuming you can reproduce the issue when using the default config. I have tried to recreate the scenario with the conditions you listed and unforunately, I still cannot reproduce.

Huh, let me try again with the minimal config

**Comment by famiu** [comment]
> > @famiu I am assuming you can reproduce the issue when using the default config. I have tried to recreate the scenario with the conditions you listed and unforunately, I still cannot reproduce.
> 
> Huh, let me try again but with the minimal init

Huh... the issue doesn't occur with the minimal config

**Comment by famiu** [comment]
Wonder if I should close this issue since I'm sure it's being caused by another plugin now or if I should wait until I find the culprit

**Comment by ram02z** [comment]
Hmm, I would probably close this. Was the reason for the splash screen bug due to how string highlights were handled by feline? I still don't get why `colors.bg` was working but `colors.fg` wasn't.

**Comment by famiu** [comment]
> Hmm, I would probably close this. Was the reason for the splash screen bug due to how string highlights were handled by feline? I still don't get why `colors.bg` was working but `colors.fg` wasn't.

Honestly, I still have no idea what caused it, but fixing string highlights seemed to fix it for whatever reason, even though the default config wasn't using any string highlights

**Comment by famiu** [comment]
> > Hmm, I would probably close this. Was the reason for the splash screen bug due to how string highlights were handled by feline? I still don't get why `colors.bg` was working but `colors.fg` wasn't.
> 
> Honestly, I still have no idea what caused it, but fixing string highlights seemed to fix it for whatever reason, even though the default config wasn't using any string highlights

And even if it is caused by another plugin, the fact that it only happens under those 3 extremely specific conditions is suspicious.

**Comment by famiu** [comment]
Either way I've decided to close this now since it's confirmed to be the fault of another plugin or a setting that causes it. Might reopen if it somehow still ends up being caused by Feline
